codyze-core/src/main/kotlin/de/fraunhofer/aisec/codyze/Sarif.kt

```diff
@@ -0,0 +1,216 @@
+ *
+ */
+package de.fraunhofer.aisec.codyze
+
+import de.fraunhofer.aisec.cpg.graph.Node
+import de.fraunhofer.aisec.cpg.graph.declarations.Declaration
+import de.fraunhofer.aisec.cpg.graph.declarations.FieldDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.NamespaceDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.ParameterDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.VariableDeclaration
+import de.fraunhofer.aisec.cpg.graph.firstParentOrNull
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Literal
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Reference
+import de.fraunhofer.aisec.cpg.query.QueryTree
+import io.github.detekt.sarif4k.*
+
+/**
+ * Converts a [QueryTree] to a list of [Result]s. This expects that the query tree is of type
+ * [Boolean] and that the [QueryTree.children] represent the individual findings.
+ */
+fun QueryTree<Boolean>.toSarif(ruleID: String): List<Result> {
+    return this.children.map {
+        Result(
+            ruleID = ruleID,
+            message =
+                if (this.value) {
+                    Message(text = "Query was successful")
+                } else {
+                    Message(text = "Query failed")
+                },
+            level =
+                if (this.value) {
+                    Level.None
+                } else {
+                    Level.Error
+                },
+            kind =
+                if (this.value) {
+                    ResultKind.Pass
+                } else {
+                    ResultKind.Fail
+                },
+            locations = listOfNotNull(it.node?.toSarifLocation()),
+            stacks = it.node?.toSarifCallStack(),
+            codeFlows =
+                it.children.mapNotNull { child ->
+                    if (child.value is List<*>) {
+                        CodeFlow(
+                            threadFlows =
+                                listOf(
+                                    ThreadFlow(
+                                        message = Message(text = "Thread flow"),
+                                        locations =
+                                            (child.value as List<*>)
+                                                .filterIsInstance<Node>()
+                                                .toSarifThreadFlowLocation(),
+                                    )
+                                )
+                        )
+                    } else {
+                        null
+                    }
+                },
+        )
+    }
+}
+
+/**
+ * Converts a [Node] into a [Message] for SARIF output. Currently, this is a short representation of
+ * the node type and node. In the future, we want to include a brief description of any eventual
+ * overlay nodes that better describe the node semantically.
+ */
+private fun Node?.toSarifMessage(): Message? {
```

Comment:
We shouldn't need the nullable return type `Message?`. We're always creating a new `Message()`.

Reply:
ah which we probably shouldn't. We should return null if the node is null